homeassistant/components/sleepiq/__init__.py

```diff
@@ -1,110 +1,199 @@
-        return "SleepNumber {} {} {}".format(
-            self.bed.name, self.side.sleeper.first_name, self._name
+
+class SleepIQDataUpdateCoordinator(DataUpdateCoordinator[dict[str, dict]]):
+    """SleepIQ data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        *,
+        client: Sleepyq,
+        update_interval: timedelta,
+        username: str,
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass, _LOGGER, name=f"{username}@SleepIQ", update_interval=update_interval
+        )
+        self.client = client
+
+    async def async_update_config(self, config_entry: ConfigEntry) -> None:
+        """Handle config update."""
+        self.update_interval = timedelta(
+            seconds=config_entry.options[CONF_SCAN_INTERVAL]
         )
 
-    def update(self):
-        """Get the latest data from SleepIQ and updates the states."""
-        # Call the API for new sleepiq data. Each sensor will re-trigger this
-        # same exact call, but that's fine. We cache results for a short period
-        # of time to prevent hitting API limits.
-        self.sleepiq_data.update()
+    async def _async_update_data(self) -> dict[str, dict]:
+        return await self.hass.async_add_executor_job(self.update_data)
+
+    def update_data(self) -> dict[str, dict]:
+        """Get latest data from the client."""
+        data = {}
+        for bed in self.client.beds_with_sleeper_status():
+            data[bed.bed_id] = {BED: bed}
+
+        return data
+
+
+@dataclasses.dataclass
+class SleepIQHassData:
+    """Home Assistant SleepIQ runtime data."""
+
+    coordinators: dict[str, SleepIQDataUpdateCoordinator] = dataclasses.field(
+        default_factory=dict
+    )
+
+
+class SleepIQSensor(CoordinatorEntity):
+    """Implementation of a SleepIQ sensor."""
 
-        self.bed = self.sleepiq_data.beds[self._bed_id]
-        self.side = getattr(self.bed, self._side)
+    _attr_icon = ICON_OCCUPIED
+
+    def __init__(
+        self,
+        coordinator: SleepIQDataUpdateCoordinator,
+        bed_id: str,
+        side: str,
+        name: str,
+    ) -> None:
+        """Initialize the SleepIQ side entity."""
+        super().__init__(coordinator)
+        self.bed_id = bed_id
+        self.side = side
+
+        self._async_update_attrs()
+
+        self._attr_name = f"SleepNumber {self._bed.name} {self._side.sleeper.first_name} {SENSOR_TYPES[name]}"
+        self._attr_unique_id = f"{self.bed_id}_{self._side.sleeper.first_name}_{name}"
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._async_update_attrs()
+        super()._handle_coordinator_update()
+
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Update sensor attributes."""
+        self._bed = self.coordinator.data[self.bed_id][BED]
+        self._side = getattr(self._bed, self.side)
```

Comment:
Looks like there is `self.side` and `self._side`?

Reply:
Yeah, one is a string (left/right), the other is the actual SideStatus object from sleepyq. I'll figure out better names for those.